Add help text to contest form

RagnarGrootKoerkamp opened the pull request:
Remaining questions:
- what is the recommended set of settings to play with a contest without it being visible to anybody else (only jury/admins, not teams/public).
- what to put for `start time enabled`
- is `enabled: false` any different from it not being active?

Fix #911

meisterT commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -39,28 +41,34 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'Yes' => true,
                 'No' => false,
             ],
+            'help' => 'Only John knows',
```
Disable to delay the contest start. Enable again after setting a new start time.

RagnarGrootKoerkamp replied:
What effect has `delay the contest start` on the GUI? Will the coundown in the top right just disappear?

meisterT replied:
Yes, it will display "countdown paused" or a similar message.

meisterT commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -39,28 +41,34 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'Yes' => true,
                 'No' => false,
             ],
+            'help' => 'Enables the balloons page. (What is the advantage of disabling it?)',
```
>  (What is the advantage of disabling it?)

This was introduced by @nickygerritsen in 1f3d6d9181ec33127186e52cbba9e12b43ac16e2 but I don't remember why exactly. For this PR, let's just remove this part of the comment.

nickygerritsen replied:
If you use an external tool to process balloons this gives a minor speed improvement but i I don’t mind to drop it since it’s not that useful

thijskh replied:
Not sure what you mean with external tool but if you disable this, the balloons table will not be filled so you can also not use an external tool I guess. I think it's indeed some kind of optimization that we do not fill another table if you don't do anything with it?

nickygerritsen replied:
I mean the cds. It can also generate balloons

meisterT reviewed (changes requested):
See https://github.com/DOMjudge/domjudge/blob/master/webapp/src/Service/ImportExportService.php#L151 for the settings that we typically enable when playing around with a contest - we nowadays import this typically with the https://github.com/DOMjudge/domjudge/blob/master/misc-tools/import-contest.sh script.

A disabled contest doesn't automatically get active even when activatetime passes.

thijskh commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -28,9 +28,11 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         $builder->add('name', TextType::class);
         $builder->add('activatetimeString', TextType::class, [
             'label' => 'Activate time',
+            'help' => 'Time when the contest becomes visible for teams to join. Must be in the past to enable judging of jury submissions.',
```
Is the second sentence true? Or should it be "to enable submission of jury submissions"? Honest question, do not know.

RagnarGrootKoerkamp replied:
I don't know either. Figured making a potentially wrong statement is the best way to find out :P

meisterT replied:
Indeed, you can't submit as jury before activate time. But if there are submissions (since you moved activatetime after submitting), we'll keep judging.

thijskh commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -39,28 +41,34 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         ]);
         $builder->add('endtimeString', TextType::class, [
             'label' => 'End time',
+            'help' => 'Time when the contest ends.',
```
+ "(Submissions will be accepted but shown as 'too-late' and not counted towards the score)"

RagnarGrootKoerkamp commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -25,12 +25,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
         $builder->add('shortname', TextType::class);
-        $builder->add('name', TextType::class);
+        $builder->add('name', TextType::class, [
+            'help' => 'Contest name as shown in the top right.'
```
Also added this one here.

thijskh replied:
I think this helptext applies to 'shortname', and 'name' is 'Contest name in full as shown on the scoreboard'

RagnarGrootKoerkamp commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -77,6 +88,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'Yes' => true,
                 'No' => false,
             ],
+            'help' => 'When true, any logged in team can join this contest, and the teams/categories listed below are added by default. When false, only the teams/categories listed below will participate.',
```
Can someone confirm this assumption is indeed correct?

nickygerritsen replied:
Nope when true the other fields are hidden (or should be) and everyone automatically is part of the contest. When false only the teams / cats you select are part. We don’t have a way to join as a team

RagnarGrootKoerkamp replied:
Ah right. They are hidden indeed.

nickygerritsen commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -25,12 +25,16 @@ public function buildForm(FormBuilderInterface $builder, array $options)
+        ]);
         $builder->add('activatetimeString', TextType::class, [
             'label' => 'Activate time',
+            'help' => 'Time when the contest becomes visible for teams to join. Must be in the past to enable submission of jury submissions.',
```
Instead of:

> Time when the contest becomes visible for teams to join

isn't

> Time when the contest becomes visible for teams

Since there is no actual joining.

RagnarGrootKoerkamp commented:
Still need to test that this text actually shows up nicely on the webpage. What's the easiest way to do that?

nickygerritsen commented:
> Still need to test that this text actually shows up nicely on the webpage. What's the easiest way to do that?

Checkout code, docker-compose up (maybe twice if it errors), http://localhost:12345 and log in with admin and the password from etc/initial_admin_password in the docker container

thijskh commented:
![image](https://user-images.githubusercontent.com/3808792/98814047-70cff980-2425-11eb-804a-30364b62a00a.png)

RagnarGrootKoerkamp commented:
Hmm, is the `Yes`/`No` button text weird in combination with the `true`/`false` description text?

thijskh commented:
Agreed, I would use "enabled"/"disabled"

RagnarGrootKoerkamp commented:
Rewrote `yes/no` to `enabled/disabled`, and changed the form titles such that `enabled/disabled` is a sensible answer (instead of `yes/no`). This may be a bit too much though. Personally I like the look of the `yes/no` buttons slightly better because they look more distinct.

Also removed the manual line wrapping from the help text in the top right, and prevented text wrapping of the time spec.

![img-2020-11-11-143342](https://user-images.githubusercontent.com/6233682/98818054-526cfc80-242b-11eb-874e-2bb070890f45.png)
![img-2020-11-11-143358](https://user-images.githubusercontent.com/6233682/98818063-54cf5680-242b-11eb-99b5-e055869f252e.png)

thijskh commented:
Sorry, I already fixed the wrapping in master. So you can rebase on that.

Enabled/disabled as labels is indeed too much. Also it is redundant in places like: "Enable public scoreboard? Enable/Disable."

RagnarGrootKoerkamp commented:
Ok, back to `Yes`/`No` button labels, but `enabled`/`disabled`  (instead of `true`/`false`) in the title and descriptions.

![img-2020-11-11-145312](https://user-images.githubusercontent.com/6233682/98819905-c0b2be80-242d-11eb-9f0e-5472cee8d37f.png)

thijskh reviewed (approved):
I am good with it but would appreciate another review

meisterT commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -24,59 +24,75 @@ class ContestType extends AbstractExternalIdEntityType
                 'No' => false,
             ],
+            'help' => 'Disable to delay the contest start and stop the countdown. Enable again after setting a new start time.',
+            # What happens when the countdown is disabled, but the start time is in the past/time 'crosses' the start time?
```
In that case, we still do the same and say "start delayed" and disallow looking at problems/submissions:
![image](https://user-images.githubusercontent.com/418721/98850097-8872a680-2454-11eb-9351-70e8ef91c5ea.png)

We may want to do just forbid this (in a separate commit), but for now just remove this line.

meisterT commented on webapp/src/Form/Type/ContestType.php:
```diff
@@ -24,59 +24,75 @@ class ContestType extends AbstractExternalIdEntityType
         ]);
         $builder->add('endtimeString', TextType::class, [
             'label' => 'End time',
+            'help' => 'Time when the contest ends. Submissions made after this time will be accepted but shown as \'too-late\' and not counted towards the score.',
```
- "accepted **and judged**"
- "but shown **(to teams and public)**"

meisterT reviewed (commented):
Two nits

meisterT commented:
Thanks!
